src/sam/functions/api/get_dmps/app.rb

```diff
@@ -36,31 +46,81 @@ def self.process(event:, context:)
-                                                                          params['owner_org_ror'].nil? &&
-                                                                          params['modification_day'].nil?
-
       _set_env(logger:)
-      logger.info(message: "DMP ID Search Criteria: #{params}") if logger.respond_to?(:debug)
+      logger&.info(message: "DMP ID Search Criteria: #{params}")
 
       # Fail if the Provenance could not be loaded
       claim = event.fetch('requestContext', {}).fetch('authorizer', {})['claims']
       provenance = Uc3DmpProvenance::Finder.from_lambda_cotext(identity: claim, logger:)
-      return _respond(status: 403, errors: Uc3DmpId::MSG_DMP_FORBIDDEN, event:) if provenance.nil?
+      return _respond(status: 403, errors: Uc3DmpId::Helper::MSG_DMP_FORBIDDEN, event:) if provenance.nil?
+
+      # Security check. If the Provenance is not allowed to see ALL orgs and the
+      #                 requested org is not in the list of their allowed orgs, then
+      #                 remove the requested org from the params
+      unless provenance.fetch('org_access_level', '').downcase != 'all'
+        rors = provenance.fetch('ror_list', []).map { |id| id.gsub('https://ror.org/', '') }
+        # Reply with a 403 if the requested Org is not in the list of approved RORs
+        if !params['org'].nil? && !rors.include?(params['org'])
+          errors = "Invalid ROR. You may make requests for the following ROR ids: #{provenance.fetch('ror_list', [])}"
+          return _respond(status: 403, errors:, event:) if provenance.nil?
+
+        elsif params['org'].nil?
+          # If they did not specify an Org, allow all of the valid ones
+          params['org'] = rors.join('|')
+        end
+      end
 
       resp = Uc3DmpId::Finder.search_dmps(args: params, logger:)
-      return _respond(status: 400, errors: Uc3DmpId::Helper::MSG_DMP_NO_DMP_ID) if resp.nil?
-      return _respond(status: 404, errors: Uc3DmpId::Helper::MSG_DMP_NOT_FOUND) if resp.empty?
+      dmps = resp.is_a?(Array) ? resp : []
+      logger&.debug(message: 'Search returned the following index records:', details: dmps)
+
+      # Perfom search operations
+      term = params.fetch('search', '').to_s.strip.downcase
+      unless term.blank?
+        logger&.debug(message: "Searching results for #{term}")
+        dmps = dmps.select do |dmp|
+          dmp['title'].to_s.downcase.include?(term) || dmp['abstract'].to_s.downcase.include?(term)
+        end
+      end
+
+      # Handle sort
+      col = params['sort'].to_s.downcase
+      dir = params['sort_dir'].to_s.downcase
+      sort = SORT_OPTIONS.include?(col) ? col : DEFAULT_SORT_OPTION
+      sort_dir = SORT_DIRECTIONS.include?(dir) ? dir : DEFAULT_SORT_DIR
+      logger&.debug(message: "Sorting results: #{col}, #{dir}")
+      dmps = dmps.sort do |a, b|
+        sort_dir == 'desc' ? b[sort] <=> a[sort] : a[sort] <=> b[sort]
```

Comment:
L93: Since you already have 'desc' saved in the DEFAULT_SORT_DIR variable, you could use it in this line.....or will it be too confusing?

Reply:
yes 💯